### PR TITLE
net: Exit early on zero gateway provided

### PR DESCRIPTION
**Pull request by finikorg**
In a case gateway is not set drop packet early otherwise bogus
ARPs for 0.0.0.0 are sent.

...
[net/arp] [DBG] find_entry: (0xa8006720): dst 0.0.0.0
[net/arp] [DBG] find_entry: (0xa8006720): [0] iface 0xa800cd80 dst
0.0.0.0 ll 00:00:00:00:00:00 pending 0xa800a7c0
[net/arp] [DBG] find_entry: (0xa8006720): ARP already pending to
0.0.0.0 ll 00:00:00:00:00:00
[net/arp] [DBG] net_arp_prepare: (0xa8006720): Resending ARP
0xa800a380
...

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

### CODE REVIEW

**Review by pfalcon — CHANGES_REQUESTED** [review]
My initial (local) patch for https://github.com/zephyrproject-rtos/zephyr/pull/1508 looked exactly like that. But that means that application (system) behavior changes depending on whether debugging is enabled. Even without bringing that to the discussion on the mailing list, it's clear that we don't want that.

Neither I wanted to make that being checked unconditionally - during my RFC discussion on the ML, the most common criticism was that "logging bloats binaries". Well, extra checks bloat binaries.

The solution I proposed in my RFC is: error/warning is enabled by default. Then, a user will see an error message, and can decide what was wrong:

* If th egateway setting was missing, it can be added.
* Otherwise, if app sent packets beyond the local net, while it shouldn't, that can be fixed.

Then, all-time runtime check is not required at all.

So, in that vein, let me be -1 on this.

**Comment by finikorg** [comment]
@pfalcon my bad, I will reformat the patch to remove debug condition check

**Comment by pfalcon** [comment]
@finikorg : Please also consider 2nd part of my message, where I argue that such a check isn't that much useful (vs adding extra code size) ;-).

**Comment by finikorg** [comment]
@pfalcon so what is the problem? Enable that debug option by default, that's it.

**Comment by pfalcon** [comment]
@finikorg : I argue that with logging enabled by default, extra check for gateway not being set is not needed (because a user will see the issue and fix it).

**Comment by finikorg** [comment]
@pfalcon I can remove this check in the following patch since this is separate change (and it may need to be agreed separately, maybe it should be done consistently with the other code)

**Comment by jukkar** [comment]
What about just having

```
if (net_is_ipv4_addr_unspecified(addr)) {
	NET_ERR("Gateway not set for iface %p",
				net_pkt_iface(pkt));

	return NULL;
}
```

Currently default log level is 1 so the error print would be printed anyway. The check for ARP debug is useless in this case.

**Comment by finikorg** [comment]
@jukkar that was change in the second patch, I can squash them

**Comment by jukkar** [comment]
This looks now a sane thing to do. @pfalcon are you ok with this?

**Comment by pfalcon** [comment]
I didn't receive a response to my argument (that *this* patch is not needed at all)/. I'd expect a response like: "No, I still think that we should have this patch, even though it unconditionally increases the code size, while not really needed in well-working, well-debugged application". Anyway, that's all for my RFC https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-September/008114.html . And when I submit a patch to enable error/warning logging by default, please don't shoot it down with "it increases binary size" arguments. It does, but other things increase it too. And there're more important things that binary size, and it's reversible (unlike patches like this).

Otherwise I don't have anything against this patch per se, just treat it as an "evaluation usecase". (Sorry to @finikorg if it went confusing.)

**Comment by finikorg** [comment]
@pfalcon Sorry I do not get what you mean. The patch is about return early on missing gateway, preventing from bogus ARP packets being sent.

**Comment by pfalcon** [comment]
@finikorg : As you ask, let me try to explain again: with the error logging enabled (like it should be), a user will get an error when the gateway is missing. It's not much important then that a bogus ARP packet is being sent - there're a bigger indication of misconfiguration (namely, the error message). A user then will fix it, and the condition this patch checks for will never be true in a well-configured, well-debugged application. Then arguably, such a patch is not much needed either (that's why I didn't include such an early return in my patch to add the logging).

**Comment by finikorg** [comment]
@pfalcon For many uses we do not need gateway, it is enough to have common network.

**Comment by pfalcon** [comment]
@finikorg : And this point was discussed above: if you don't need a gateway, then your app should not be sending stuff outside local network. If it does, again you will see the error and can fix your app.

Anyway, this whole discussion isn't as much of this change, than about how the developers actually treat the matter of code size growth. It's just very small case, which is easy to consider from all sides. I consider my point proven - people don't care about the code size as the first priority. So, I hope it will be considered the same way when time to enable error logging by default comes.

**Review by pfalcon — DISMISSED** [review]

**Comment by pfalcon** [comment]
Ah, I see, my review was still in "changes requested", fixed that now. (github doesn't allow to be +0, only +1 or -1).

**Comment by jukkar** [comment]
> And when I submit a patch to enable error/warning logging by default, please don't shoot it down with "it increases binary size" arguments.

I have never objected setting default log level to warning (currently it is error, and beside there are not many warnings in net stack atm). I know it is increasing binary size and that is ok in this case. What I objected in the other PR, was to enable debug logging via the global flag which makes it actually hard to debug things (because of extensive output).

**Review by jukkar — APPROVED** [review]

**Review by jukkar — CHANGES_REQUESTED** [review]
Actually noticed that the commit msg is missing "net: arp: ..." subject.

**Comment by pfalcon** [comment]
@jukkar: Thanks for the clarification, and for merging the first steps so far. I'm doing homework on the next steps (goes slow as I'm on semi-vacation, then a conference).

**Review by jukkar — APPROVED** [review]
